src/dune_pkg/opam_solver.ml

```diff
@@ -518,15 +492,22 @@ let opam_package_to_lock_file_pkg
     ; extra_sources
     }
   in
-  (* This will collect all the atoms from the package's dependency formula regardless of conditions *)
   let deps =
-    Context_for_dune.filter_deps context_for_dune opam_package opam_file.depends
-    |> OpamFormula.fold_right
-         (fun acc (name, _condition) ->
-           if OpamPackage.Name.Set.mem name all_package_names then name :: acc else acc)
-         []
-    |> List.map ~f:(fun name ->
-      Loc.none, Package_name.of_string (OpamPackage.Name.to_string name))
+    match
+      Resolve_opam_formula.filtered_formula_to_package_names
+        ~with_test:false
+        solver_env
+        version_by_package_name
+        opam_file.depends
+    with
+    | Ok dep_package_names ->
+      List.map dep_package_names ~f:(fun package_name -> Loc.none, package_name)
+    | Error (`Formula_could_not_be_satisfied hints) ->
+      Code_error.raise
```

Comment:
Isn't this possible? why is it a code error?

Reply:
It shouldn't be possible. This runs immediately following the solver and the solution should contain packages that satisfy the dependency formulae of each package in the solution.